### PR TITLE
xdstp: CDS/RDS/SRDS/EDS support added.

### PR DESCRIPTION
**Pull request by htuch**
This adds the foundations of support for glob collections in CDS/SRDS
(in addition to LDS from #14311) and singletons in RDS/EDS.

xdstp:// URLs are not normalized across the internal resource graph yet,
so this is only safe to use for sorted context parameters (followup PRs
planned on this).

Part of #11264.

Risk level: Low (opt-in)
Testing: ADS/SRDS integration tests added, various unit tests in
  //test/common/config/...

Signed-off-by: Harvey Tuch <htuch@google.com>

### CODE REVIEW

File: source/common/config/watch_map.cc [hunk 1]
```diff
@@ -13,25 +13,9 @@ namespace Config {
...
-    // Replace resource name component with glob for purpose of matching.
-    const auto pos = resource.id().find_last_of('/');
-    resource.set_id(pos == std::string::npos ? "*" : resource.id().substr(0, pos) + "/*");
-    XdsResourceIdentifier::EncodeOptions encode_options;
-    encode_options.sort_context_params_ = true;
-    return XdsResourceIdentifier::encodeUrn(resource, encode_options);
-  }
-  // In the non-xdstp:// legacy case for VHDS, we simple remove the last / component. E.g.
-  // www.foo.com/bar becomes www.foo.com.
+  // We simple remove the last / component. E.g. www.foo.com/bar becomes www.foo.com.
```
Comment (adisuissa) [inline source/common/config/watch_map.cc:16]: Nit: *simply

File: source/common/router/scoped_rds.cc [hunk 2]
```diff
@@ -103,16 +104,28 @@ ScopedRdsConfigSubscription::ScopedRdsConfigSubscription(
...
-  subscription_ =
-      factory_context.clusterManager().subscriptionFactory().subscriptionFromConfigSource(
-          scoped_rds.scoped_rds_config_source(), Grpc::Common::typeUrl(resource_name), *scope_,
-          *this, resource_decoder_, false);
+  std::unique_ptr<xds::core::v3::ResourceLocator> srds_resources_locator;
+  if (!scoped_rds.srds_resources_locator().empty()) {
+    srds_resources_locator = std::make_unique<xds::core::v3::ResourceLocator>(
+        Envoy::Config::XdsResourceIdentifier::decodeUrl(scoped_rds.srds_resources_locator()));
+  }
+  if (srds_resources_locator == nullptr) {
```
Comment (adisuissa) [inline source/common/router/scoped_rds.cc:118]: What's the advantage of using the `srds_resources_locator` pointer? Shouldn't this be `if (scoped_rds.srds_resources_locator().empty())` instead?

**Review by adisuissa — COMMENTED** [review]
Overall looks good, thanks!

File: source/common/config/new_grpc_mux_impl.cc [hunk 3]
```diff
@@ -145,58 +145,53 @@ GrpcMuxWatchPtr NewGrpcMuxImpl::addWatch(const std::string& type_url,
...
-    if (absl::EndsWith(resource.id(), "/*")) {
-      auto encoded_context = XdsContextParams::encodeResource(
-          local_info_.contextProvider().nodeContext(), resource.context(), {}, {});
-      resource.mutable_context()->CopyFrom(encoded_context);
+  // We need to prepare xdstp:// resources for the transport, by normalizing and adding any extra
+  // context parameters.
+  absl::flat_hash_set<std::string> effective_resources;
+  for (const auto& resource : resources) {
+    if (XdsResourceIdentifier::hasXdsTpScheme(resource)) {
+      auto xdstp_resource = XdsResourceIdentifier::decodeUrn(*resources.begin());
```
[the remark this answers is not in the document]
Comment (htuch) [reply source/common/config/new_grpc_mux_impl.cc:180]: Good catch! Will add such tests.

File: source/common/router/scoped_rds.cc [hunk 4]
```diff
@@ -103,16 +104,26 @@ ScopedRdsConfigSubscription::ScopedRdsConfigSubscription(
...
-      factory_context.clusterManager().subscriptionFactory().subscriptionFromConfigSource(
-          scoped_rds.scoped_rds_config_source(), Grpc::Common::typeUrl(resource_name), *scope_,
-          *this, resource_decoder_, false);
+  if (scoped_rds.srds_resources_locator().empty()) {
+    subscription_ =
+        factory_context.clusterManager().subscriptionFactory().subscriptionFromConfigSource(
+            scoped_rds.scoped_rds_config_source(), Grpc::Common::typeUrl(resource_name), *scope_,
+            *this, resource_decoder_, {});
+  } else {
+    std::unique_ptr<xds::core::v3::ResourceLocator> srds_resources_locator =
```
Comment (adisuissa) [inline source/common/router/scoped_rds.cc:119]: I might be missing something, but why is heap allocation needed in this case?

File: source/common/config/new_grpc_mux_impl.cc [hunk 5]
```diff
@@ -145,58 +145,53 @@ GrpcMuxWatchPtr NewGrpcMuxImpl::addWatch(const std::string& type_url,
...
+      effective_resources.insert(XdsResourceIdentifier::encodeUrn(xdstp_resource, encode_options));
     } else {
-      // TODO(htuch): We will handle list collections here in future work.
-      NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+      effective_resources.insert(resource);
     }
   }
-  auto added_removed = sub->second->watch_map_.updateWatchInterest(
-      watch, xdstp_resources.empty() ? resources : xdstp_resources);
-  if (creating_namespace_watch && xdstp_resources.empty()) {
```
Comment (adisuissa) [inline source/common/config/new_grpc_mux_impl.cc:199]: This implies that namespace matching and xdstp are mutually exclusive. Should there be an assert somewhere? Maybe just add a comment to the struct.

**Review by adisuissa — COMMENTED** [review]
Overall looks good.
Left a couple of minor comments.

File: source/common/config/new_grpc_mux_impl.cc (hunk 5 above)
Comment (htuch) [reply source/common/config/new_grpc_mux_impl.cc:199]: Comment added; namespace matching basically should never be used outside of VHDS, and even there we will eventually support `xdstp://`.

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!

**Comment by htuch** [comment]
@mattklein123 this is ready for review, @adisuissa has done first pass and LGTMed.

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**